ibis/spark/client.py

```diff
@@ -0,0 +1,284 @@
+    pt.ShortType : dt.int16,
+}
+
+
+@dt.dtype.register(pt.DataType)
+def spark_type_to_ibis_dtype(spark_type_obj):
+    """Convert Spark SQL types to ibis types."""
+
+    if isinstance(spark_type_obj, pt.DecimalType):
+        precision = spark_type_obj.precision
+        scale = spark_type_obj.scale
+        ibis_type = dt.Decimal(precision, scale)
+    elif isinstance(spark_type_obj, pt.ArrayType):
+        value_type = dt.dtype(spark_type_obj.elementType)
+        nullable = spark_type_obj.containsNull
+        ibis_type = dt.Array(value_type, nullable)
+    elif isinstance(spark_type_obj, pt.MapType):
+        key_type = dt.dtype(spark_type_obj.keyType)
+        value_type = dt.dtype(spark_type_obj.valueType)
+        nullable = spark_type_obj.valueContainsNull
+        ibis_type = dt.Map(key_type, value_type, nullable)
+    elif isinstance(spark_type_obj, pt.StructType):
+        names = spark_type_obj.names
+        fields = spark_type_obj.fields
+        ibis_types = list(map(dt.dtype, fields))
+        ibis_type = dt.Struct(names, ibis_types)
+    else:
+        ibis_type = _DTYPE_TO_IBIS_TYPE.get(type(spark_type_obj))
+
+    return ibis_type
+
+
+@sch.infer.register(ps.sql.dataframe.DataFrame)
+def spark_dataframe_schema(df):
+    """Infer the schema of a Spark SQL `DataFrame` object."""
+    fields = OrderedDict((el.name, dt.dtype(el.dataType)) for el in df.schema)
+
+    return sch.schema(fields)
+
+
+class SparkCursor:
+    """Spark cursor.
+
+    This allows the Spark client to reuse machinery in
+    :file:`ibis/client.py`.
+
+    """
+
+    def __init__(self, query):
+        """
+
+        Construct a SparkCursor with query `query`.
+
+        Parameters
+        ----------
+        query : pyspark.sql.DataFrame
+          Contains result of query.
+
+        """
+        self.query = query
+
+    def fetchall(self):
+        """Fetch all rows."""
+        result = self.query.collect()  # blocks until finished
+        return result
+
+    @property
+    def columns(self):
+        """Return the columns of the result set."""
+        return self.query.columns
+
+    @property
+    def description(self):
+        """Get the fields of the result set's schema."""
+        return self.query.schema
+
+    def __enter__(self):
+        # For compatibility when constructed from Query.execute()
+        """No-op for compatibility.
+
+        See Also
+        --------
+        ibis.client.Query.execute
+
+        """
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """No-op for compatibility.
+
+        See Also
+        --------
+        ibis.client.Query.execute
+
+        """
+
+
+class SparkQuery(Query):
+    def __init__(self, client, ddl):
+        super().__init__(client, ddl)
+
+    def _fetch(self, cursor):
+        df = cursor.query.toPandas()  # blocks until finished
+        schema = self.schema()
+        return schema.apply_to(df)
+
+    def execute(self):
```

Comment:
This method is duplicated from `Query.execute` in `ibis/client.py`, so you can remove it.